### ISSUE TITLE
feature like "index_merge" from MySQL, only tidb can implement the optimizer

### ISSUE THREAD

**Issue by jianhaiqing** [pingcap/tidb#9019]
## Feature Request

**Is your feature request related to a problem? Please describe:**
SQL which contain OR works very slow in TIDB, but work well in MySQL;
+ select SQL looks like as follows:
```
SELECT  
  c_1, c_2, c_3
FROM  
  t_or_test  
WHERE  
  (c_1  = 'xxx' 
  OR  c_2  = 'yyy'
);
```
+ DDL   
```
CREATE TABLE `t_or_test` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `c_3` int(11) NOT NULL DEFAULT '0',
  `c_2` char(120) NOT NULL DEFAULT '',
  `c_1` char(40) NOT NULL DEFAULT '',
  `c_hai` varchar(20) NOT NULL DEFAULT 'ghost',
  PRIMARY KEY (`id`),
  KEY `idx_c1` (`c_1`),
  KEY `idx_c2` (`c_2`)
) ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8mb4
```

**Describe the feature you'd like:**
TIDB can optimize the sql, working like 'union all';

**Teachability, Documentation, Adoption, Migration Strategy:**

**Comment by XuHuaiyu** [comment]
@jianhaiqing Thanks for your feedback.

**Comment by morgo** [comment]
Confirming output of EXPLAIN: 
```
mysql> explain analyze SELECT     c_1, c_2, c_3 FROM     t_or_test   WHERE     (c_1  = 'xxx'    OR  c_2  = 'yyy' );
+-----------------------+----------+------+--------------------------------------------------------------------+----------------------------------+
| id                    | count    | task | operator info                                                      | execution info                   |
+-----------------------+----------+------+--------------------------------------------------------------------+----------------------------------+
| Projection_4          | 8000.00  | root | test.t_or_test.c_1, test.t_or_test.c_2, test.t_or_test.c_3         | time:861.245µs, loops:1, rows:0  |
| └─TableReader_7       | 8000.00  | root | data:Selection_6                                                   | time:814.558µs, loops:1, rows:0  |
|   └─Selection_6       | 8000.00  | cop  | or(eq(test.t_or_test.c_1, "xxx"), eq(test.t_or_test.c_2, "yyy"))   |                                  |
|     └─TableScan_5     | 10000.00 | cop  | table:t_or_test, range:[-inf,+inf], keep order:false, stats:pseudo | time:0s, loops:0, rows:0         |
+-----------------------+----------+------+--------------------------------------------------------------------+----------------------------------+
4 rows in set (0.00 sec)
```

**Comment by jianhaiqing** [comment]
Yes. you can reproduce the output from mysql.5.7.

**Comment by zz-jason** [comment]
@jianhaiqing Thanks for your report, we are going to support this feature in the near future. 😄

**Comment by zz-jason** [comment]
confirmed that this issue is fixed since TiDB release 4.0. You can enable the index merge feature and see the improvement:

```sql
TiDB(root@127.0.0.1:test) > set @@tidb_enable_index_merge = 1;
Query OK, 0 rows affected (0.01 sec)

TiDB(root@127.0.0.1:test) > desc SELECT     c_1, c_2, c_3 FROM     t_or_test   WHERE     (c_1  = 'xxx'    OR  c_2  = 'yyy' );
+---------------------------+---------+-----------+---------------------------------------------------------------------------------+
| id                        | estRows | task      | operator info                                                                   |
+---------------------------+---------+-----------+---------------------------------------------------------------------------------+
| Projection_4              | 8000.00 | root      | test.t_or_test.c_1, test.t_or_test.c_2, test.t_or_test.c_3                      |
| └─IndexMerge_11           | 20.00   | root      |                                                                                 |
|   ├─IndexRangeScan_8      | 10.00   | cop[tikv] | table:t_or_test, index:c_1, range:["xxx","xxx"], keep order:false, stats:pseudo |
|   ├─IndexRangeScan_9      | 10.00   | cop[tikv] | table:t_or_test, index:c_2, range:["yyy","yyy"], keep order:false, stats:pseudo |
|   └─TableRowIDScan_10     | 20.00   | cop[tikv] | table:t_or_test, keep order:false, stats:pseudo                                 |
+---------------------------+---------+-----------+---------------------------------------------------------------------------------+
5 rows in set (0.01 sec)
```
